### ISSUE TITLE
Early microcode loading

### ISSUE THREAD

**Issue by 01micko** [puppylinux-woof-CE/woof-CE#2344]
In [this post](https://github.com/puppylinux-woof-CE/woof-CE/pull/2316#issuecomment-864349820) in #2316 I performed a little experiment to see if early microcode loading would work in Puppy with our `initrd.[gx]z` and it worked! (Tested Intel only - will test AMD later).

**EDIT**: new code posted

What I propose is an entirely optional mechanism to concatenate the generated `ucode.cpio` archive to the current `initrd.gz` at the woof level and in userspace. It uses [this guide](https://www.kernel.org/doc/Documentation/x86/microcode.txt).

In woof, it can be done quite easily if you have built a kernel and added firmware from [kernel.org](https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/) for AMD but Intel microcode needs to be grabbed from [Intel's github repo](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/releases).
~In userspace, it is easier to grab the AMD microcode from [googlesource (direct link to tar archive)](https://kernel.googlesource.com/pub/scm/linux/kernel/git/firmware/linux-firmware.git/+archive/refs/heads/main/amd-ucode.tar.gz)~ ~(yes I can hear the sighs :scream: "No, not Google!" ).~ EDIT: **now getting code from kernel.org** :tada: 
It is a mirror of the kernel.org firmware repo and will be current. The advantage is a direct link to the tar archive. _Note: it may be possible to do a sparse checkout from kernel.org with the right pattern_
The Intel files are updated regularly.

The userspace implementation could be part of *Boot Manager* and give the options:
- AMD, Intel or both.
- whether to backup the stock `initrd.gz` or not

Anyway, here is a script, useful for userspace and woof that just gets the microcode (AMD, Intel or both) and builds the `ucode.cpio` in `/tmp`.

```sh
#!/bin/sh

# gets the latest intel and amd microcode for cpu bugs and builds
# a cpio archive containing this code in the correct hierarchy
# for early loading of microcode.
# refer https://www.kernel.org/doc/Documentation/x86/microcode.txt
# GPLv2

# the intent is for the script output to be fairly quiet

set -e
[ -e /etc/rc.d/PUPSTATE ] && . /etc/rc.d/PUPSTATE
rm -rf 
ver=0.1
PROG="${0##*\/}"

usage_() {
	echo "$PROG-$ver"
	echo
	echo "$PROG -h|--help : show this message"
	echo "$PROG amd : build AuthenticAMD.bin microcode"
	echo "$PROG intel : build GenuineIntel.bin microcode"
	echo "$PROG with no arguments build both AMD and Intel microcode"
	exit 2 
}

TMPDIR=/tmp/microcode
rm -rf $TMPDIR
PKG_INTEL=/initrd/mnt/tmpfs/tmp/microcode/kernel/x86/microcode
PKG_AMD=
AMDM=1
INTM=1
LABEL=''
intel_func() {
	# download - intel github
	SRC_URL_INTEL="https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/releases"
	PRE_URL_INTEL=$(curl -s $SRC_URL_INTEL | grep -om1 '\/intel.*tar\.gz')
	PKG_INTEL=${PRE_URL_INTEL##*\/}
	URL_INTEL="https://github.com/${PRE_URL_INTEL}"
	wget -q $URL_INTEL || return 1
	tar axf $PKG_INTEL || return 1
}

amd_func() {
	# refer https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/amd-ucode
	SRC_URL_AMD="https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/plain/amd-ucode/"
	mkdir -p amd-ucode
	curl -s https://git.kernel.org/pub/scm/linux/kernel/git/firmware/linux-firmware.git/tree/amd-ucode | grep -o 'plain.*bin' | sed 's/plain.*ucode\///' | sort | uniq |\
	while read AMD_UCODE
	do
		wget -q $SRC_URL_AMD/$AMD_UCODE -P amd-ucode || return 1
	done
}

case $1 in 
	*-h*|h*)usage_             ;;
	amd|AMD)AMDM=0; LABEL=$1       ;;
	intelIntel)INTM=0; LABEL=$1     ;;
	'')AMDM=0; INTM=0          ;;
	*)echo "wrong arg"; exit 1 ;;
esac
mkdir -p $TMPDIR
cd $TMPDIR
TGTDIR=kernel/x86/microcode
mkdir -p $TGTDIR
if [ $INTM -eq 0 ];then
	intel_func || exit 1
	INTEL_DIR=$(ls | grep "Intel\-Linux")
	cat $INTEL_DIR/intel-ucode/* > $TGTDIR/GenuineIntel.bin
	rm -rf $INTEL_DIR *gz # clean up to build ucode.cpio
fi
if [ $AMDM -eq 0 ];then
	amd_func || exit 1
	cat amd-ucode/*.bin > $TGTDIR/AuthenticAMD.bin
	rm -rf amd-ucode/ # clean up to build ucode.cpio
fi

# generate ucode initrd
rm -f /tmp/ucode.* # clean up old file
find . | cpio -o -H newc > ../ucode.cpio || exit 1 # can't be compressed
cd - >/dev/null 2>&1
if [ -e /tmp/ucode.cpio ];then
	echo "$LABEL ucode.cpio is in /tmp"
else
	echo "$LABEL ucode.cpio failed to be generated"
	exit 1
fi
# copy to puphome
if [ -e /mnt/home/$PSUBDIR ];then
	# force copy to overwrite old version
	cp -af /tmp/ucode.cpio /mnt/home/$PSUBDIR/ || exit 1
else
	exit 1
fi
	
exit 0

```

Any thoughts @dimkr @peabee @rizalmart @gyrog @zigbert ?

**Comment by jamesbond3142** [comment]
1. Nice code, it's a keeper.

2. In Fatdog, we do early-microcode loading. We don't use late-loading because new microcode _could_ cause CPU feature change. In early loading, the kernel can re-init itself and adapts accordingly, in late loading, all the kernel initialisation is already done so if there are changes, it _might_ not be able to adapt properly (and develop difficult-to-debug problems later).

3. We don't use concatenated initrd as well because it makes further editing of initrd difficult. We just embed the microcode as part of initrd building process, the usual initrd editing is easy (and to update microcode, just need to unpack, delete the "kernel" directory, copy over the new initrd, and re-pack). 

4. With certain bootloaders, you can specify multiple initrds; the first one can be the usual one and the other one can be the microcode.cpio. It works as well. I remember grub4dos and syslinux can do this, but grub legacy could not (apparently grub2 gained the ability on version 2.04 back in 2018), and IIRC refind doesn't support it as well. That is why we don't use this method (not everybody runs a supported bootloader), even though conceptually this would be the best solution.

**Comment by 01micko** [comment]
> 1. Nice code, it's a keeper.

Thanks!

> 2. In Fatdog, we do early-microcode loading. We don't use late-loading because new microcode _could_ cause CPU feature change. In early loading, the kernel can re-init itself and adapts accordingly, in late loading, all the kernel initialisation is already done so if there are changes, it _might_ not be able to adapt properly (and develop difficult-to-debug problems later).

Yes I read about CPU feature change and thought that early loading is the correct solution for Puppy. Also, we don't need to carry those heavy intel microcode files in /lib/firmware by default.

> 3. We don't use concatenated initrd as well because it makes further editing of initrd difficult. We just embed the microcode as part of initrd building process, the usual initrd editing is easy (and to update microcode, just need to unpack, delete the "kernel" directory, copy over the new initrd, and re-pack).

My first thought was that FatDog uses huge initrd and that could be problematic but I see that editing the initrd is probably the best solution, which is certainly food for thought going forward. 

In my opinion, I think this *shouldn't* be part of woof-ce by default (but can easily be made optional) and I certainly lean toward the userspace implementation. Bloat is the main concern but on the other hand should a small amount of bloat compromise security? More food for thought. :thinking: 

> 4. With certain bootloaders, you can specify multiple initrds; the first one can be the usual one and the other one can be the microcode.cpio. It works as well. I remember grub4dos and syslinux can do this, but grub legacy could not (apparently grub2 gained the ability on version 2.04 back in 2018), and IIRC refind doesn't support it as well. That is why we don't use this method (not everybody runs a supported bootloader), even though conceptually this would be the best solution.

Really, specifying more than one initrd makes it messy for automated solutions like `frugalpup` (by @gyrog ) and can fully understand why you reject this solution given the variety of bootloaders available..

My mind on this is a bit clearer and I will certainly do some more testing and evaluation before I decide on the path we should traverse :smiley: 

Cheers, and thanks for your valuable input.

**Comment by rizalmart** [comment]
That's the problem of early loading it requires modifying bootloader configuration or even repacking of initrd. That's why there is late-loading of microcode on my pull request. Unless create a workaround in order to implement early loading of microcode dynamically for example a dummy ucode.cpio file

**Comment by gyrog** [comment]
1. I don't understand why folk continualy think that adding a a file to initrd.gz is difficult, you just need the right tools, (blowing own trumpet), that's why I produced a utility called 'initrdfile', which provides functions to 'get|put|delete' a file in an initrd.gz, I've published this in the old forum some time ago. But I found that I updated the 'init' script so freuently that I have unpublished sripts  'getinit' and 'putinit' (these use 'initrdfile').  With a few simple sripts, updating the 'initrd.gz' with 'ucode.cpio' should not be a problem at all.
2. (Still blowing my own trumpet), if you use "FrugalPup" to install Puppy then you are booting with grub2. The non-uefi grub2 in frugalpul is already at 2.04, so that might be a means of testng the multiple initrd's. The uefi grub2 is still v 2.03 with thanks to @jamesbond3142, so it might take a repeat effort to test multiple initrd's in uefi boots. Although I was hoping to put off a request for new uefi grub2 files until 2.06 is released. FrugalPup has the capability to easily relpace or append to a 'grub.cfg' without disturbing anything else. These days, the only editing I do to my 'grub.cfg' is to move boot-entries around, and delete some.
3. I have no problem working on updating the 'init' script if it is required, or utility scrpts to make it easy to use 'ucode.cpio' with 'initrd.gz', or updating FrugalPup to deal with multiple 'initrds'. I just need some information on how this works, and therefore what needs to be done.
4. I like the idea of the multiple 'initrd.gz' files, having the micro-code updated before even 'init' runs sounds good. But  how do you tell grub2 to use them?

To me, "modifying bootloader configuration or even repacking of initrd", either of these is not a big deal.

**Comment by Butterfly** [comment]
>     4. I like the idea of the multiple 'initrd.gz' files, having the micro-code updated before even 'init' runs sounds good. But  how do you tell grub2 to use them?

Hi, s'me, the perennial lurker who may have misunderstood so my apologies in advance. Regarding the multiple initrd.gz files, I use `lilo` just enough to be dangerous.

One line that you can add to each `lilo` block in /etc/lilo.conf is `append`. I've not had to adjust that yet so I'm not sure of how far it can be pushed. I *do* know that multiple different things can be attached that are not addressed in the other lines, e.g. `image` and `initrd`.

And now that I type that, I wonder if it's possible to present multiple options in that single `initrd` line. I *have* seen that suggested for something that may have been that specific line. I remember trying it, but it failed. My educated guess is that it was User error in that case.

Just thinking out loud in hopes maybe there's a potentially comparable feature available for grub2. :guide_dog:

**Comment by gyrog** [comment]
Is this what we are talking about?

Quote from GNU GRUB Manual 2.06
"Command: initrd file [file …]
    Load, in order, all initial ramdisks for a Linux kernel image, and set the appropriate parameters in the Linux setup area in memory. This may only be used after the linux command (see linux) has been run."

Note: I just found out that Grub 2.06 has been released.

**Comment by Butterfly** [comment]
> Is this what we are talking about?
> 
> Quote from GNU GRUB Manual 2.06
> "Command: initrd file [file …]
> Load, in order, all initial ramdisks for a Linux kernel image, and set the appropriate parameters in the Linux setup area in memory. This may only be used after the linux command (see linux) has been run."

Was refreshing bird feeders which loosened up my brain. What about something under probably most likely `/etc/default/grub`? Or is that what's being described. I don't have grub2 installed to peek at what's already there (it doesn't like my GPT setup for some reason). To me, that directory has always felt like a more "static" place where Developers stash necessary recommendations that Users can then alter if need be.

There's also `/etc/grub.d` that can be customized. The last time I used it, that directory was starting to change some, seemingly for the better regarding serious User boot personalizations.

**Comment by dimkr** [comment]
For my use cases, which don't have an initrd at all (very similar to a "full installation"), I prefer microcode loading early in rc.sysinit, and I do think it should be an integral part of woof-CE, like other firmware/boot loader/etc' related stuff.

In other cases, I think yes - microcode should be loaded as early as possible and that means initrd. Slightly less-early loading can be implemented by putting the microcode files in the main SFS, then trying to load them both in initrd (before switching to the real init) and in rc.sysinit (for "full installations").

Another option worth looking into is integrating an initrd with the microcode into the kernel image, and loading Puppy's initrd as a second initrd on top. I don't know if this is possible, both technically and legally, but a solution like this would solve the problem for all installation types.

**Comment by jamesbond3142** [comment]
> The uefi grub2 is still v 2.03 with thanks to @jamesbond3142, so it might take a repeat effort to test multiple initrd's in uefi boots.

It works there too.

According to commit grub2-git commit c30f378e00d3fde86de9fc48f4ec5a3d2a62fdea, apparently the feature was there since 2012, but was only announced in v2.04 in 2018, and the docs on how to use it (which, as you found it, @gyrog, is "initrd file1 file2 ...") was only updated in the above commit, dated Aug 8 2020 ...

> Another option worth looking into is integrating an initrd with the microcode into the kernel image,
and loading Puppy's initrd as a second initrd on top. I don't know if this is possible, both technically and legally, but a solution like this would solve the problem for all installation types.

This works, but don't ask me about the legality :)
The only downside is that it ties the microcode to the kernel at the kernel build-time, as it's not easy to replace an initrd embedded inside the kernel.

**Comment by CollaboratorGCM** [comment]
In JamesBond's #3, 01Mick's response, and GyroG's findings: I think ALL agree at a "reimage" of init for boot. Now, its a mere mechanics as has been started by the opening stanza, here.

I think its a great approach to settle in on usings the world's GRUB2 latest release and the settling in on GRUB2-EFI as PUP's direction will make PUP/DOG distros one of the best(if not THE BEST) distro for system startups in the Linux arena.

I have tested multiple "GRUB2 ISO file Boot Launchers", all of them produce a boot to modern PUP/DOG desktops. This is NOT the case for all distros, but, so far, in my testing PUP-DOG is recognized, presented, and launched without issues in their use. This has been tested in UEFI configurations with/without Secure boot as well as in BIOS configurations.

And, although this is a Shutdown (end of session) issue, **Persistence** in both PUP/DOG runnings are unaffected by boots from either ISO file base or Frugal base with EXACT operations of the system upon reboot, no matter which method the users selects for maintaining the base system. In my case, this single feature for ISO file boots has shortened "considerably" the turn-around for testing new PUP/DOGs in the kennel on either bare-metal or QEMU.

As small a change this has been to booting (namely GRUB2-EFI + loopback), it has been a productivity gain thanks to your attention and efforts. I cannot thank you enough.

**Comment by gyrog** [comment]
1. Does this have to be done every boot, or just once?
2. If it's every boot, then we could simply use a modified "initrd" line in 'grub.cfg' with microcode as first initrd file, simple. With a script like 'bootentry' that generates a bootentry to "stdout",  'grub.cfg' can easily be re-built. And it can be tested now, by manually editing a 'grub'cfg'.
3. If it's only once, then we could look at getting conditional code in 'grub.cfg' to test for the presence of the microde initrd, and use the appropriate existing initrd line. Again with a script that can produce the new boot-entries that include that code, not a big deal.
4. If current booting does not use an 'initrd' line, then would it work with adding an 'initrd' line with only the microcode initrd?
5. Sorry, off-topic, but using this new grub2 facility, we could split the normal 'initrd.gz' into 2,  1 that contains only DISTRO_SPECS, and 1 that contains all the rest. The same 2nd initrd could be used for all Puppies, only the DISTRO_SPECS one is different.

**Comment by jamesbond3142** [comment]
> Cheers, and thanks for your valuable input.

@01micko No worries, happy to share :+1:

@gyrog 
> 1. Does this have to be done every boot, or just once?

Every boot.

> 2. If it's every boot, then we could simply use a modified "initrd" line in 'grub.cfg' with microcode as first initrd file, simple. With a script like 'bootentry' that generates a bootentry to "stdout",  'grub.cfg' can easily be re-built. And it can be tested now, by manually editing a 'grub'cfg'.

Correct

> 4. If current booting does not use an 'initrd' line, then would it work with adding an 'initrd' line with only the microcode initrd?

You're thinking in the case of full install? The answer is yes.

**Comment by peabee** [comment]
@ozsouth has been providing microcode downloads and howto for some time.....
https://forum.puppylinux.com/viewtopic.php?f=104&t=1327

**Comment by gyrog** [comment]
FYI,
The following code works with 2 initrd files, or 1 initrd file:
+++++++++++
  if [ -e /puppy/upuphh/distro-initrd.gz ]; then
    echo "Loading distro-initrd.gz and initrd.gz"
    initrd /puppy/upuphh/distro-initrd.gz /puppy/upuphh/initrd.gz
  else
    echo "Loading initrd.gz"
    initrd /puppy/upuphh/initrd.gz
  fi
+++++++++++
We can easily produce a solution that has an optional initrd file containing microcode.
If it exists in the same directory as the kernel, then it gets loaded first.

It also means that we could split the current initrd.gz into 2, one that is the same for all puppies and the other containing only DISTRO_SPECS, exactly as I did here.
Of course having 2 optional initrd files would make the code a bit more complicated, but then there would be no real need for folk to add this stuff in manually.

Later edit: It works with both uefi and non-uefi versions of grub2.

**Comment by 01micko** [comment]
> @ozsouth has been providing microcode downloads and howto for some time.....
> https://forum.puppylinux.com/viewtopic.php?f=104&t=1327

But no AuthenticAMD.bin , not at least in the latest.

**Comment by 01micko** [comment]
Hmmm.. 

> Slightly less-early loading can be implemented by putting the microcode files in the main SFS

The files can be downloaded according to processor (at least with Intel). All 132 of them are just bloat when you only need 1 file.

```
#!/bin/bash

# gets the correct microcode for your processor for late loading
# GPLv2
# bashism

# refer https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files
# get the ucode version
while read var 
do 
	if [ "${var:0:10}" = 'cpu family' ];then
		CPUFAM=${var##* }
	fi
 	if [ "${var:0:5}" = 'model' ];then
		MODEL=${var##* }
	fi
	if [ "${var:0:8}" = 'stepping' ];then
		STEP=${var##* }
	fi	
done <<< $(cat /proc/cpuinfo | grep -E 'cpu family|model|stepping' | grep -v 'model name' | sort -u )

# convert to hex
c=1
for i in $CPUFAM  $MODEL $STEP ; do 
	if [ ${#i} -lt 2 ] ;then
		case $c in
			1)CHAR1=0;;
			2)CHAR2=0;;
			3)CHAR3=0;;
		esac
	else
		case $c in
			1)CHAR1= ;;
			2)CHAR2= ;;
			3)CHAR3= ;;
		esac
	fi
	[ $c -gt 3 ] && break # failed
	c=$(($c + 1))
done
INT_UCODE=$(printf "${CHAR1}%x-${CHAR2}%x-${CHAR3}%x\n" $CPUFAM $MODEL $STEP)

MURL=https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/raw/main/intel-ucode

# download to /lib/firmware - must be root
mkdir -p /lib/firmware/intel-ucode 
if wget -nc -q $MURL/$INT_UCODE -P /lib/firmware/intel-ucode ;then
	echo "$INT_UCODE installed succesfully"
	exit 0
else
	echo "$INT_UCODE failed to download"
	exit 1
fi
```

**Comment by gyrog** [comment]
It's best if the downloaded microcode file is not "integrated" into any other Puppy file.
Not good to tie the release of new microcode for a users CPU, with a Puppy release.
If it's a standalone file in the same directory as the kernel file  then it's easily accessible to grub2.
If it's a standalone file in the same directory as the puppy...sfs then it's easily accessible to 'init' and a running Puppy.

**Comment by 01micko** [comment]
@gyrog that's why my OTF version, it takes only a second on a fast connection and because 1 file is only a matter of kilobytes - and mine is one of the larger blobs @ 100 kB. Of course internet connection must be up.

EDIT: we can also grok out the microcode file and version from the [intel releases](https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/releases) page.
```sh
curl -s https://github.com/intel/Intel-Linux-Processor-Microcode-Data-Files/releases | grep -A2 -m1 "${INT_UCODE}" | sed -e 's%<*td.*\">%%' -e 's%\/.*$%%' -e 's%<%%' # gets microcode file \n old-version \n new-version \n
```
We can store the current version in a config file
eg:
```sh
CURRENT_VER=read a b micver rest <<< $(od -t x4 /lib/firmware/intel-ucode/$INT_UCODE | head -n1)
echo $CURRENT_VER > $HOME/.intel-ucode-version
```

**Comment by 01micko** [comment]
@gyrog @dimkr @jamesbond3142 @peabee @rizalmart 

We have some design questions to answer.

1. Do we go with 2 initrd archives or do we copy the microcode to the existing initrd? This of course depends on if the bootloader can handle it. Either way they are upgradable.
2. Do we carry the microcode files in the `fdrv` or do we use something similar to my OTF code? (supports intel only ATM but AMD shouldn't be too difficult). In the `fdrv` they are (can be?) there _but_ they'll always be as old as the kernel package used. OTF guarantees the latest; this is more particular to late loading, which concerns full installs (rare), possibly save back to optical installs (pupmode 77) and @dimkr 's `frugalify` type installs that don't use an initrd (although the `ucode.cpio` should/could be used anyways)
3. Should we even support late loading? I can't see any harm with the right conditional routines.
4. Conditional implementation of the above.

Thoughts?

**Comment by rizalmart** [comment]
> @gyrog @dimkr @jamesbond3142 @peabee @rizalmart
> 
> We have some design questions to answer.
> 
> 1. Do we go with 2 initrd archives or do we copy the microcode to the existing initrd? This of course depends on if the bootloader can handle it. Either way they are upgradable.
> 2. Do we carry the microcode files in the `fdrv` or do we use something similar to my OTF code? (supports intel only ATM but AMD shouldn't be too difficult). In the `fdrv` they are (can be?) there _but_ they'll always be as old as the kernel package used. OTF guarantees the latest; this is more particular to late loading, which concerns full installs (rare), possibly save back to optical installs (pupmode 77) and @dimkr 's `frugalify` type installs that don't use an initrd (although the `ucode.cpio` should/could be used anyways)
> 3. Should we even support late loading? I can't see any harm with the right conditional routines.
> 4. Conditional implementation of the above.
> 
> Thoughts?

My thoughts:
1. Don't ever touch the initrd. It make things more complicated. Microcode on separate initrd is good but you need to deal with configurations of different bootloaders
2. It depends but microcode frequently updates by cpu manufacturers
3. YES, late loading of microcode after switch_root for minimal vulnerabilities is safe bet.
4. Microcode loading is optional but gains advantage when enable it

**Comment by gyrog** [comment]
"Don't ever touch the initrd. It make things more complicated." I could not disagree too strongly to this statement.
I've done an awful lot of changing the initrd.gz over the last few years, without any problems with the process.

Now to the questions:

4  The simplest form of a flag for loading or not loading microcode is the existence of the microcode file in the same directory as the kernel.
If someone wants to do microcode loading for a particular Puppy they obtain the latest version of the microcode suitable for their CPU and place it in the install directory.
This provides the use with a choice todo or not todo, and if todo, with which kernels todo it.
Obviously there would need to be a "get-microcode" utility.

1  The sticking point here is, can we get a CD to boot with the 2 initrd file model?
If booting from a CD with a multisession capable DVD, you get one chance to write the microcode file to the DVD as a new file,
to update to a new version of microcode you would need to burn a new DVD, assuming that no microcode file is included in a release.
We could use a 2 initrd model, with a merged initrd.gz as a backup plan for bootloaders that don't support multiple initrd files.
But this would only work if Puppy is installed on a simple re-writable device like a usb stick, it would not work with a DVD, since it already has an initrd.gz file.
The thing I like about this is that it does now require changing Puppy itself, just a few utilities, (which could be portable-applications),
so it should be easily implemented on current Puppies, provided the kernel being used supports it.

2  Embedding the microcode file in any Puppy file only makes it more difficult for users to introduce the latest microcode as soon as it is released.
That's one reason why I prefer the 2 initrd file model. New microcode, drop a new microcode file in the install directory and reboot, or a symbolic link to the microcode file.

3  The only advantage I can see here, is that we have a ready made solution available, well I assume the pull-request is that.

**Comment by jamesbond3142** [comment]
My two cents.

> 1. Do we go with 2 initrd archives or do we copy the microcode to the existing initrd? This of course depends on if the bootloader can handle it. Either way they are upgradable.

As said earlier, my ideal preferred solution is to have 2 initrd. Fatdog doesn't do that way because rEFInd doesn't support it (and we use rEFInd in some scenarios), but if this is not a limitation for you, then best to go for it.

Here is my suggestion: 
- Have microcode inside the 1st initrd. This ensures that some updates are already available even if we only load a single initrd.
- Optionally have the 2nd initrd containing updates / newer version of the microcode. The microcode from the 2nd initrd will overwrite the first one, and that's the one that gets used.
(Note: ordering is important. Main initrd with puppy stuff (and older microcode) is the first one, 2nd initrd containing optional updates only).

It doesn't matter the size of the microcode in initrd. The ISO size will be a bit bigger, but RAM usage will be unaffected, because after switch_root this extra baggage is gone and not kept in RAM (the kernel will automatically cache only the microcode part that it actually uses - which is very small). This is unlike if you put it in the SFS, which will be kept in RAM for as long as the OS is running.

In case of full install, skip the 1st initrd and just use the 2nd initrd. All other full-install configuration can remain the same.

> 
> 2. Do we carry the microcode files in the `fdrv` or do we use something similar to my OTF code? (supports intel only ATM but AMD shouldn't be too difficult). In the `fdrv` they are (can be?) there _but_ they'll always be as old as the kernel package used. OTF guarantees the latest; this is more particular to late loading, which concerns full installs (rare), possibly save back to optical installs (pupmode 77) and @dimkr 's `frugalify` type installs that don't use an initrd (although the `ucode.cpio` should/could be used anyways)
> 

Don't bother. The case of _frugalify_ could be treated as if it is a full-install, as above.

> 3. Should we even support late loading? I can't see any harm with the right conditional routines.

Don't bother. 

> 
> 4. Conditional implementation of the above.
> 

I'm not sure which implementation you want to make conditional, but with the 2-initrd mode, it can be made "conditional" just by excluding / not loading the 2nd initrd. All you need is to have two boot configurations (one that loads microcode, and one that doesn't).

cheers!

**Comment by 01micko** [comment]
Quite successfully booting my AMD box with latest `ucode.cpio` using @gyrog 's grub2 stanza with appropriate editing

```
menuentry "Puppy slacko64 8.2.1" {
  insmod ext2
  search --no-floppy --fs-uuid --set=root fac65aa2-84bf-4dde-949c-75fefeac5e29
  echo "Loading vmlinuz"
  linux /pups/slacko64-8.2.1/vmlinuz pmedia=atahd pdrv=NVPUP psubdir=/pups/slacko64-8.2.1 pfix=fsck,fsckp,trim TZ=AEST-10
  if [ -e /pups/slacko64-8.2.1/ucode.cpio ]; then
    echo "Loading ucode.cpio and initrd.gz"
    initrd /pups/slacko64-8.2.1/ucode.cpio /pups/slacko64-8.2.1/initrd.gz
  else
    echo "Loading initrd.gz"
    initrd /pups/slacko64-8.2.1/initrd.gz
  fi
}
```

And works

```
~$ dmesg|grep -i microcode
[    0.755914] microcode: microcode updated early to new patch_level=0x0800820d
[    0.755933] microcode: CPU0: patch_level=0x0800820d
[    0.755938] microcode: CPU1: patch_level=0x0800820d
[    0.755940] microcode: CPU2: patch_level=0x0800820d
[    0.755946] microcode: CPU3: patch_level=0x0800820d
[    0.755951] microcode: CPU4: patch_level=0x0800820d
[    0.755956] microcode: CPU5: patch_level=0x0800820d
[    0.755961] microcode: CPU6: patch_level=0x0800820d
[    0.755968] microcode: CPU7: patch_level=0x0800820d
[    0.755974] microcode: CPU8: patch_level=0x0800820d
[    0.755980] microcode: CPU9: patch_level=0x0800820d
[    0.755985] microcode: CPU10: patch_level=0x0800820d
[    0.755992] microcode: CPU11: patch_level=0x0800820d
[    0.755995] microcode: Microcode Update Driver: v2.2.
[    8.211711] [drm] Loading oland Microcode
```

**Comment by gyrog** [comment]
Is the microcode initrd always going to be called 'ucode.cpio'?

**Comment by 01micko** [comment]
> Is the microcode initrd always going to be called 'ucode.cpio'?

Yes.

That's what the kernel doc says. It can be named anything.  But my test scripts have it hard coded. So I'll keep it, seems self explanatory to a ~geek~ developer I suppose.

EDIT: also note it can't be compressed with gz,xz or zst.

**Comment by jamesbond3142** [comment]
> > Is the microcode initrd always going to be called 'ucode.cpio'?
> 
> Yes.
> 
> That's what the kernel doc says. It can be named anything. But my test scripts have it hard coded. So I'll keep it, seems self explanatory to a ~geek~ developer I suppose.
> 

I'd suggest you attach a timestamp as part of the filename so everyone knows when the microcode was pulled out. We called ours "microcode-update-20200905.cpio" for example.

**Comment by jamesbond3142** [comment]
@01micko - your updated script in the first post has ```rm -rf```. Are you sure it's supposed to be there?

**Comment by 01micko** [comment]
> @01micko - your updated script in the first post has `rm -rf`. Are you sure it's supposed to be there?

No. Must have been a copy/paste typo. Thanks to the beauty of open source it is spotted early! Thanks!

> > > Is the microcode initrd always going to be called 'ucode.cpio'?
> > 
> > 
> > Yes.
> > That's what the kernel doc says. It can be named anything. But my test scripts have it hard coded. So I'll keep it, seems self explanatory to a ~geek~ developer I suppose.
> 
> I'd suggest you attach a timestamp as part of the filename so everyone knows when the microcode was pulled out. We called ours "microcode-update-20200905.cpio" for example.

I'm thinking more along the lines of a config file; I can grok out the timeline for intel from the github repo 'release'  page but amd I may have to record when the files were downloaded.

**Comment by gyrog** [comment]
Please keep the filename of the microcode initrd file constant.
We do not have "bash" or even "ash" scripting code available to detect the presence of this file,
just the scripting facilities available in grub2.
Of course this may be a lot more powerful than I give it credit for, but then documentation is rather minimal.

**Comment by 01micko** [comment]
> Please keep the filename of the microcode initrd file constant.
> We do not have "bash" or even "ash" scripting code available to detect the presence of this file,
> just the scripting facilities available in grub2.
> Of course this may be a lot more powerful than I give it credit for, but then documentation is rather minimal.

My intention is to keep it constant, and record the timestamp in a config file.

**Comment by dimkr** [comment]
Got early loading to work with efilinux, by putting ucode.cpio (built separately) in EFI/BOOT/ucode.cpio and modifying EFI/BOOT/efilinux.cfg from:

```
-f 0:\EFI\BOOT\vmlinuz root=PARTLABEL=dpupos_root init=/init rootfstype=ext4 rootwait rw
```

to:

```
-f 0:\EFI\BOOT\vmlinuz root=PARTLABEL=dpupos_root initrd=0:\EFI\BOOT\ucode.cpio init=/init rootfstype=ext4 rootwait rw
```

```
[    0.000000] microcode: microcode updated early to revision 0x2f, date = 2019-02-17
[    2.490052] microcode: sig=0x206a7, pf=0x10, revision=0x2f
[    2.490189] microcode: Microcode Update Driver: v2.2.
```
